twenty-first/src/util_types/algebraic_hasher.rs

```diff
@@ -78,6 +88,50 @@ pub trait AlgebraicHasher: SpongeHasher {
 
         Digest::new((&produce[..DIGEST_LENGTH]).try_into().unwrap())
     }
+
+    /// Produce `num_indices` random integer values in the range `[0, upper_bound)`.
+    ///
+    /// - The randomness depends on `state`.
+    /// - `upper_bound` must be a power of 2.
+    ///
+    /// If `num_indices` is not divisible by `RATE`, spill the remaining elements of the last `squeeze`.
+    fn sample_indices(
+        state: &mut Self::SpongeState,
+        upper_bound: u32,
+        num_indices: usize,
+    ) -> Vec<u32> {
+        assert!(is_power_of_two(upper_bound));
+        let num_squeezes = (num_indices + Self::RATE - 1) / Self::RATE;
+        (0..num_squeezes)
+            .flat_map(|_| Self::squeeze(state))
+            .take(num_indices)
+            .map(|b| (b.value() % upper_bound as u64) as u32)
+            .collect()
+    }
+
+    /// Produce `num_elements` random [XFieldElement] values.
+    ///
+    /// - The randomness depends on `state`.
+    ///
+    /// Since [RATE] is not divisible by [EXTENSION_DEGREE], produce as many [XFieldElement] per
+    /// `squeeze` as possible, and spill the remaining element(s). This causes some internal
+    /// fragmentation, but it greatly simplifies building [AlgebraicHasher::sample_xfield()] on
+    /// Triton VM.
+    fn sample_scalars(state: &mut Self::SpongeState, num_elements: usize) -> Vec<XFieldElement> {
+        // let num_squeezes = (num_elements * EXTENSION_DEGREE + Self::RATE - 1) / Self::RATE;
+        let num_squeezes = num_elements * EXTENSION_DEGREE / Self::RATE;
```

Comment:
Because of spill, I don't think this is correct. Consider:

```rust
assert_eq!(3, EXTENSION_DEGREE);
assert_eq!(10, Self::RATE);
let num_elements = 10;
let num_squeezes = num_elements * EXTENSION_DEGREE / Self::RATE; // 3, but should be 4
```

```suggestion
        const XFES_PER_SQUEEZE = Self::Rate / EXTENSION_DEGREE;
        let num_squeezes = (num_elements + XFES_PER_SQUEEZE - 1) / XFES_PER_SQUEEZE;
```

Picking up above example:

```rust
const XFES_PER_SQUEEZE = Self::Rate / EXTENSION_DEGREE; // 3
let num_squeezes = (num_elements + XFES_PER_SQUEEZE - 1) / XFES_PER_SQUEEZE; // 4
```

Consider collecting into a vector and asserting the length of the vector before returning.

Reply:
An assertion as described can also happen in some test, but that test should be more thorough than the currently existing one.